File: src/services/snippet-manager.ts
import { Snippet } from '../models/snippet';
import { SecureStorage } from './secure-storage';
import { ipcMain } from 'electron';

export class SnippetManager {
  private secureStorage: SecureStorage;

  constructor(secureStorage: SecureStorage) {
    this.secureStorage = secureStorage;
  }

  async runSnippet(snippet: Snippet, terminalId: string): Promise<boolean> {
    try {
      // Send the command to the terminal
      const windows = require('electron').BrowserWindow.getAllWindows();
      for (const window of windows) {
        if (!window.isDestroyed()) {
          window.webContents.send('terminal:data', {
            terminalId,
            data: snippet.command + '\n',
          });
        }
      }

      return true;
    } catch (error) {
      console.error('Failed to run snippet:', error);
      return false;
    }
  }

  async runSnippetsOnConnect(
    snippetIds: string[],
    terminalId: string
  ): Promise<void> {
    try {
      // Get all snippets
      const snippets = await Promise.all(
        snippetIds.map((id) => this.secureStorage.getSnippet(id))
      );

      // Filter out null values and snippets that shouldn't run on connect
      const snippetsToRun = snippets.filter(
        (s): s is Snippet => s !== null && s.runOnConnect
      );

      // Run each snippet with a small delay between them
      for (const snippet of snippetsToRun) {
        await this.runSnippet(snippet, terminalId);
        // Small delay to ensure commands execute in order
        await new Promise((resolve) => setTimeout(resolve, 100));
      }
    } catch (error) {
      console.error('Failed to run snippets on connect:', error);
    }
  }
}

File: src-tauri/src/services/snippet_manager.rs
use crate::models::snippet::Snippet;
use std::sync::{Arc, Mutex};
use tauri::Window;

pub struct SnippetManager {
  snippets: Arc<Mutex<Vec<Snippet>>>,
}

impl SnippetManager {
  pub fn new() -> Self {
    SnippetManager {
      snippets: Arc::new(Mutex::new(Vec::new())),
    }
  }

  pub fn run_snippet(
    &self,
    snippet: &Snippet,
    terminal_id: &str,
    window: Window,
  ) -> Result<bool, String> {
    match window.emit(
      "terminal:data",
      &serde_json::json!({
          "terminalId": terminal_id,
          "data": format!("{}\n", snippet.command),
      }),
    ) {
      Ok(_) => Ok(true),
      Err(err) => {
        eprintln!("Failed to run snippet: {}", err);
        Err(format!("Failed to run snippet: {}", err))
      }
    }
  }

  pub async fn run_snippets_on_connect(
    &self,
    snippet_ids: Vec<String>,
    terminal_id: &str,
    window: Window,
  ) -> Result<(), String> {
    let snippets = self.snippets.lock().unwrap();

    // Filter snippets that should run on connect
    let snippets_to_run: Vec<&Snippet> = snippets
      .iter()
      .filter(|s| snippet_ids.contains(&s.id) && s.run_on_connect)
      .collect();

    // Run each snippet with a small delay between them
    for snippet in snippets_to_run {
      self.run_snippet(snippet, terminal_id, window.clone())?;
      // Small delay to ensure commands execute in order
      tokio::time::sleep(tokio::time::Duration::from_millis(100)).await;
    }

    Ok(())
  }
}
